Adding GUC parameter to define the behavior of CATC…

Ahmad87Homs opened the pull request:
### Description
Handling Unmapped PostgreSQL Errors in T-SQL TRY...CATCH

**Issue**
Babelfish mimics SQL behavior for control flow and transaction state whenever possible. When Babelfish encounters an error, it returns an error code similar to the SQL Server error code. If Babelfish can't map the error to a SQL Server code, it returns a fixed error code (33557097) and terminates the entire batch without respecting a TRY... CATCH construct if it exists.

for more details: follow the open discussion https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/4034

**Fix**
Adding GUC params "**babelfishpg_tsql.catch_unmapped_errors**" and let's user define the session behavior, terminate and ignore catch block or handle catch block.

- By default the behviour of Babelfish is to terminat and igonre CATCH block if Error is un_mapped,

> catch_unmapped_errors = false;

- to disable defualt behavior, and handle CATCH block, the disable param sets to true

> catch_unmapped_errors = true;



### Issues Resolved

- Handling Catch Block in case unmapped Error id coughed.
- Allow user to define behavior. 

### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Ahmad87Homs committed:
BUG2025080536805: Adding GUC parameter to define the behavior of CATCH block with unmapped Error occurred

Ahmad87Homs committed:
updating Tests and Expected values

Ahmad87Homs committed:
rename tests

Ahmad87Homs committed:
Disabling interrupt on logging termination msg

Ahmad87Homs committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs committed:
Merge branch 'BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs committed:
Update guc.c

Formatting guc.c file

Ahmad87Homs committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs commented:
Hi @tanscorpio7 , thanks for the clarification earlier. I’ve created a pull request implementing a configurable parameter that allows users to define the behavior for unmapped errors.

Could you please take a look at the PR and share your feedback?

tanscorpio7 commented:
Thanks @Ahmad87Homs, I'll take a look next week.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/guc.c:
```diff
@@ -1244,6 +1245,16 @@ define_custom_variables(void)
+			                false,
+			                PGC_USERSET,
+			                GUC_NOT_IN_SAMPLE | GUC_DISALLOW_IN_FILE | GUC_DISALLOW_IN_AUTO_FILE,
+			                NULL, NULL, NULL);				
```
Let's use this name `babelfishpg_tsql.escape_hatch_catch_unmapped_errors`

Ahmad87Homs replied:
Renamed the GUC to babelfishpg_tsql.escape_hatch_catch_unmapped_errors done,

tanscorpio7 commented on contrib/babelfishpg_tsql/src/guc.c:
```diff
@@ -1719,3 +1730,13 @@ metadata_inconsistency_check_enabled(void)
+	*/
+	bool unmapped_error_termination_disabled(void){
+		return pltsql_disable_unmapped_error_termination;
+	}
```
We can directly read the global variable, no need for a get_function()

Ahmad87Homs replied:
Removed the unnecessary getter and now read the global variable directly,

tanscorpio7 commented on contrib/babelfishpg_tsql/src/iterative_exec.c:
```diff
@@ -1614,9 +1614,19 @@ exec_stmt_iterative(PLtsql_execstate *estate, ExecCodes *exec_codes, ExecConfig_
+						* Wrap elog() with HOLD_INTERRUPTS()/RESUME_INTERRUPTS() for consistency.
+						* This ensures no risk with logging during cleanup.
+						*/
+						HOLD_INTERRUPTS();
```
HOLD_INTERRUPTS is only needed when logging inside a PG_CATCH block

Ahmad87Homs replied:
Dropped the HOLD_INTERRUPTS() call where it wasn’t needed

Ahmad87Homs committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs committed:
Rename GUC to name babelfishpg_tsql.escape_hatch_catch_unmapped_errors, and remove unnecessary stuff

Ahmad87Homs committed:
Extern pltsql_enable_ownership_chaining variable

Ahmad87Homs committed:
Update expected values

Ahmad87Homs commented:
Thanks a lot, @tanscorpio7 🙏

I’ve applied the suggested changes:

Renamed the GUC to babelfishpg_tsql.escape_hatch_catch_unmapped_errors
Removed the getter function and now directly access the global variable
Removed the unnecessary HOLD_INTERRUPTS() call

Bur after renaming the GUC to escape_hatch_catch_unmapped_errors, I noticed some tests started failing. They use babelfishpg_tsql.escape_hatch% and now match this new variable, which causes unexpected test results.

Could you advise how you’d like to proceed?
Keep it in the escape_hatch_* family and update the test expectations accordingly, or
Rename it to something like babelfishpg_tsql.catch_unmapped_errors to avoid matching the existing wildcard?

tanscorpio7 commented on contrib/babelfishpg_tsql/src/guc.c:
```diff
@@ -1718,4 +1729,4 @@ bool
 {
 	return enable_metadata_inconsistency_check;
-}
+}
```
Let's avoid any unrelated changes.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/iterative_exec.c:
```diff
@@ -1614,8 +1614,16 @@ exec_stmt_iterative(PLtsql_execstate *estate, ExecCodes *exec_codes, ExecConfig_
+						/*
+						* Wrap elog() with HOLD_INTERRUPTS()/RESUME_INTERRUPTS() for consistency.
+						* This ensures no risk with logging during cleanup.
+						*/
```
This is comment is stale now.

tanscorpio7 commented:
> Rename it to something like babelfishpg_tsql.catch_unmapped_errors to avoid matching the existing wildcard?

Sounds good to me. Couple of minor comments. Thanks.

Ahmad87Homs committed:
remove stale  comments

Ahmad87Homs committed:
Reorder GUC variable definitions in guc.c

Ahmad87Homs commented:
> > Rename it to something like babelfishpg_tsql.catch_unmapped_errors to avoid matching the existing wildcard?
> 
> Sounds good to me. Couple of minor comments. Thanks.

Thanks, @tanscorpio7 

I reverted the unrelated parts and keep the changes focused. also GUC name changed to babelfishpg_tsql.catch_unmapped_errors.

Ahmad87Homs committed:
rename GUC to babelfishpg_tsql.catch_unmapped_errors

Ahmad87Homs committed:
remove extra blank line

Ahmad87Homs committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_5_X_DEV

Ahmad87Homs committed:
Removing extra blank line from guc.h

Ahmad87Homs committed:
match format of guc.h

Ahmad87Homs committed:
Fix missing newline at end of guc.h

Ahmad87Homs committed:
Fix missing newline at end of guc.h

Fix missing newline at end of guc.h

Ahmad87Homs commented:
Hi @pranavJ23  , @rohit01010 , Could you please review the PR when you have a moment? Thanks!

tanscorpio7 commented:
@Ahmad87Homs After discussing with the team, we've decided to keep the existing behavior. Adding a GUC would introduce additional error-level permutations that increase complexity. We'll revisit this proposal if we receive more requests for this configuration option.
